### PR TITLE
RightSizer controller

### PR DESCRIPTION
**Pull request by ivanfetch-fw**
## Description
Add a Kubernetes controller to maintain an Insights report when OOM-kill events are detected on Pods. Currently the report lists the owning pod-controller (Deployment), container name, and memory limit.

See also: the [pull request for the insights-agent Helm chart](https://github.com/FairwindsOps/charts/pull/615).

See the [ReadMe](https://github.com/FairwindsOps/insights-plugins/blob/right-sizer-jam/plugins/right-sizer/README.md) and [to-do](https://github.com/FairwindsOps/insights-plugins/blob/right-sizer-jam/plugins/right-sizer/TODO.md) where I've documented background, and known issues / future plans.

## Checklist

* [ ] I have updated the CHANGELOG of any plugins updated
* [ ] I have updated the version file of any plugins updated
* [ ] I have added tests wherever appropriate
* [ ] I have labeled this PR with a label per plugin updated

### CODE REVIEW

File: plugins/right-sizer/cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var opts struct {
+	Verbose int  `env:"VERBOSE" short:"v" long:"verbose" description:"Show verbose debug information"`
+	Version bool `long:"version" description:"Print version information"`
+}
+
+// VERSION represents the current version of the release.
+const VERSION = "v1.2.0"
```
Comment (rbren) [inline plugins/right-sizer/cmd/main.go:20]: I think this should be `0.1.0`.

Maybe we should remove this functionality for now - we have some automated tools for modifying `version.txt` and I think we'd forget to update this.

File: plugins/right-sizer/cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	* up causing the same channel be closed twice.
+	This created a panic when a signal (SIGTerm) was received.
+	This is now commented out, in favor of the channel being closed below via the struct member.*/
+	// util.InstallSignalHandler(stopChan)
+
+	eventGenerator := controller.NewController(stopChan)
+	util.InstallSignalHandler(eventGenerator.Stop)
+
+	http.Handle("/metrics", promhttp.Handler())
+	// addr := fmt.Sprintf("0.0.0.0:10254")
```
Comment (rbren) [inline plugins/right-sizer/cmd/main.go:41]: remove comment

File: plugins/right-sizer/cmd/main.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		printVersion()
+		return
+	}
+
+	stopChan := make(chan struct{})
+	/*ifetch: This was in the original oom-kill-event-generator code, but ends
+	* up causing the same channel be closed twice.
+	This created a panic when a signal (SIGTerm) was received.
+	This is now commented out, in favor of the channel being closed below via the struct member.*/
+	// util.InstallSignalHandler(stopChan)
```
Comment (rbren) [inline plugins/right-sizer/cmd/main.go:35]: I think you can remove these comments. Good catch!

File: plugins/right-sizer/cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	go func() { glog.Fatal(http.ListenAndServe(addr, nil)) }()
+
+	err := eventGenerator.Run()
+	if err != nil {
+		glog.Fatal(err)
+	}
+}
+
+func printVersion() {
+	fmt.Printf("kubernetes-oom-event-generator %s %s/%s %s\n", VERSION, runtime.GOOS, runtime.GOARCH, runtime.Version())
```
Comment (rbren) [inline plugins/right-sizer/cmd/main.go:52]: we should change the name here

File: plugins/right-sizer/results.schema [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema",
+    "type": "object",
+    "required": [
+    ],
+    "properties": {
+    }
```
Comment (rbren) [inline plugins/right-sizer/results.schema:7]: can we add a TODO to fill this schema out (unless I missed it)?

File: plugins/right-sizer/src/controller/controller.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+const (
+	// informerSyncMinute defines how often the cache is synced from Kubernetes
+	informerSyncMinute = 2
+	// TerminationReasonOOMKilled is the reason of a ContainerStateTerminated that reflects an OOM kill
+	TerminationReasonOOMKilled = "OOMKilled"
+)
+
+// Controller is a controller that listens on Pod changes and create Kubernetes Events
+// when a container reports it was previously killed
+type Controller struct {
```
Comment (rbren) [inline plugins/right-sizer/src/controller/controller.go:35]: there's a lot going on in this struct. I might suggest breaking it down into some sub-structs (e.g. you could group the k8s API stuff into a sub-struct).

File: plugins/right-sizer/src/controller/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
+package controller
```
Comment (rbren) [inline plugins/right-sizer/src/controller/controller_test.go:1]: Thanks for the tests!

File: plugins/right-sizer/src/controller/controller_test.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+		Reason:    reason,
+		Message:   fmt.Sprintf(messageFmt, args...),
+	})
+	return
+}
+
+func (r *dummyEventRecorder) PastEventf(_object runtime.Object, _timestamp metav1.Time, _eventtype, _reason, _messageFmt string, _args ...interface{}) {
+}
+
+func (r *dummyEventRecorder) AnnotatedEventf(_object runtime.Object, _annotations map[string]string, _eventtype, _reason, _messageFmt string, _args ...interface{}) {
```
Comment (rbren) [inline plugins/right-sizer/src/controller/controller_test.go:133]: I think these two functions are unimplemented/unused?

File: plugins/right-sizer/src/report/report.go [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+// AddOrUpdateItem accepts a RightSizerReportItem and adds or updates it
+// in the report.
+// If the item already exists, `NumOOMs` and `LastOOM` fields are updated.
+func (b *RightSizerReportBuilder) AddOrUpdateItem(newItem RightSizerReportItem) {
+	b.itemsLock.Lock()
+	defer b.itemsLock.Unlock()
+	for i, item := range b.Report.Items {
+		if item.Kind == newItem.Kind && item.ResourceNamespace == newItem.ResourceNamespace && item.ResourceName == newItem.ResourceName && item.ResourceContainer == newItem.ResourceContainer {
+			// UPdate the existing item.
```
Comment (rbren) [inline plugins/right-sizer/src/report/report.go:85]: `s/UPdate/Update/`

File: plugins/right-sizer/src/report/report.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+	}
+	return b
+}
+
+// AddOrUpdateItem accepts a RightSizerReportItem and adds or updates it
+// in the report.
+// If the item already exists, `NumOOMs` and `LastOOM` fields are updated.
+func (b *RightSizerReportBuilder) AddOrUpdateItem(newItem RightSizerReportItem) {
+	b.itemsLock.Lock()
+	defer b.itemsLock.Unlock()
```
Comment (rbren) [inline plugins/right-sizer/src/report/report.go:82]: 👍 nice job with the thread safety!

File: plugins/right-sizer/src/report/report.go [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+// RemoveOldItems removes items from the report that have a last OOMKill older than
+// RightSizerReportBuilder.tooOldAge, and returns true if any items were
+// removed.
+func (b *RightSizerReportBuilder) RemoveOldItems() (anyItemsRemoved bool) {
+	glog.V(2).Infoln("starting removing old items based on last OOM kill. . .")
+	var finalItems []RightSizerReportItem
+	b.itemsLock.Lock()
+	defer b.itemsLock.Unlock()
+	for _, item := range b.Report.Items {
+		if time.Since(item.LastOOM) >= b.tooOldAge {
```
Comment (rbren) [inline plugins/right-sizer/src/report/report.go:159]: I think `tooOldAge` will default to `0` - should we check if it's non-zero?

File: plugins/right-sizer/src/controller/controller.go (hunk 6 above)
Comment (ivanfetch-fw) [reply plugins/right-sizer/src/controller/controller.go:32]: @rbren How's this? See [this commit](https://github.com/FairwindsOps/insights-plugins/commit/6f174f8839c5b46fd9b4d62a78cc7d35644d0e88) where I broke out the client, dynamicClient, and RESTMapper as those are already instantiated and provided at the same time.

This makes the `Controller` struct a bit nicer, but also use of the client/dynamicClient/RESTMapper resources a bit more verbose E.G. `c.kubeClientResources.Client`

File: plugins/right-sizer/src/controller/controller.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+			}
+		}
+		return
+	}
+	pod, err := c.podLister.Pods(event.InvolvedObject.Namespace).Get(event.InvolvedObject.Name)
+	if err != nil {
+		glog.Errorf("Failed to retrieve pod %s/%s, due to: %v", event.InvolvedObject.Namespace, event.InvolvedObject.Name, err)
+		return
+	}
+	c.evaluatePodStatus(pod)
```
Comment (makoscafee) [inline plugins/right-sizer/src/controller/controller.go:181]: the logic above was duplicated in `evaluateEvent` and here ... makes sense to refactor it

File: plugins/right-sizer/src/report/report.go [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+// TODO: populate defaults (state ConfigMap, too-old-age) from CLI flags.
+func NewRightSizerReportBuilder(kubeClient kubernetes.Interface) *RightSizerReportBuilder {
+	b := &RightSizerReportBuilder{
+		Report:     &RightSizerReportProperties{},
+		itemsLock:  &sync.RWMutex{},
+		kubeClient: kubeClient,
+		tooOldAge:  8.64e+13, // 24 HRs
+		// tooOldAge: 6e+10, // 1 minute
+		// tooOldAge:              3e+11, // 5 minutes
+		configMapNamespaceName: "insights-agent",
```
Comment (makoscafee) [inline plugins/right-sizer/src/report/report.go:67]: Are we planning to make this configurable?

File: plugins/right-sizer/src/report/report.go [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+
+// AddOrUpdateItem accepts a RightSizerReportItem and adds or updates it
+// in the report.
+// If the item already exists, `NumOOMs` and `LastOOM` fields are updated.
+func (b *RightSizerReportBuilder) AddOrUpdateItem(newItem RightSizerReportItem) {
+	b.itemsLock.Lock()
+	defer b.itemsLock.Unlock()
+	for i, item := range b.Report.Items {
+		if item.Kind == newItem.Kind && item.ResourceNamespace == newItem.ResourceNamespace && item.ResourceName == newItem.ResourceName && item.ResourceContainer == newItem.ResourceContainer {
+			// Update the existing item.
```
Comment (makoscafee) [inline plugins/right-sizer/src/report/report.go:85]: when `ResourceVersion` is different do we still consider this as same item?

**Review by makoscafee — APPROVED** [review]
LGTM ... just had few questions.

**Review by rbren — APPROVED** [review]

**Pull request merged**